STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,440 @@
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        idle_state_id = state_id;
+        if (state_id == AUTO_PILOT) {
+            return AutoPilot;
+        }
+        else if (state_id == MANUAL_OPERATION_WAITING) {
+            return ManualControl;
+        } 
+    }
+
+    return Idle;
+}
+
+State_t StateMachineThread::AutoPilotEvent(void)
+{
+    // Set LED colour to yellow
+    SetLedColour(50.0, 50.0, 0.0);
+
+    // Send ACK on CAN when stop complete 
+    if (NewState != CurrentState) {
+        CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_STATE_CHANGE_ACK_NACK);
+        CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_STATE_ID_ACK_NACK, idle_state_id);
+        if (CANBus_put_frame(&tx_frame) != HAL_OK) { Error_Handler(); }
+    }
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    // Receive CAN frame
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        int distance_to_end = TRACK_LENGTH - global_mc_data.curr_pos;
+        if (state_id == EMERGENCY_BRAKE || state_id == SYSTEM_FAILURE) {
+            return SevereDangerFault;
+            // TODO: Talk to Ryan/Dev when to enter MinorDangerFault
+        }
+        else if (state_id == BRAKING || distance_to_end <= DISTANCE_THRESHOLD) {
+            return Idle;
+        } 
+    }
+    return AutoPilot;
+}
+
+State_t StateMachineThread::ManualControlEvent(void)
+{
+    // Set LED colour to blue
+    SetLedColour(0, 0, 50.0);
+
+    // Send ACK on CAN when stop complete
+    if (NewState != CurrentState) {
+        CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_STATE_CHANGE_ACK_NACK);
+        CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_STATE_ID_ACK_NACK, idle_state_id);
+        if (CANBus_put_frame(&tx_frame) != HAL_OK) { Error_Handler(); }
+    }
+
+    // List of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANZThread, PIDThread, VHzThread, SVPWMThread,  StateMachineThread
+
+    MeasurementsThread::resumeMeasurements();
+
+    if (!Queue_empty(&RX_QUEUE))
+    {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        run_state_id = state_id;
+        if (state_id == EMERGENCY_BRAKE || state_id == SYSTEM_FAILURE)
+        {
+            return SevereDangerFault;
+    }
+        else if (state_id == BRAKING)
+        {
+            return Idle;
+        }
+    }
+    return ManualControl;
+}
+
+State_t StateMachineThread::InitializeFaultEvent(void)
```

Comment:
Off the top of my head, a few things the MC should do in a fault (aside from what you've already put):
1. Put all the MOSFETs into HIGH-Z mode using the `drv8323` drivers
  i. The DRV drivers aren't merged in right now, so just put a comment to remind yourself to put that in once it        does get merged
2. (Probably) read the fault codes from the DRV
3. Disable the PID, SVPWM, and VHz threads

This applies to all the fault states